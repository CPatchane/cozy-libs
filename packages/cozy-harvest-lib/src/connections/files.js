const FILES_DOCTYPE = 'io.cozy.files'

/**
 * Creates a directory from a given path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}         Directory attributes
 */
export const createDirectoryByPath = async (client, path) => {
  const { data } = await client
    .collection(FILES_DOCTYPE)
    .createDirectoryByPath(path)
  return data
}

/**
 * Retrieves a directory from its path
 * @param  {Object}  client CozyClient
 * @param  {string}  path   Directory path
 * @return {Object}        Created io.cozy.files document
 */
export const statDirectoryByPath = async (client, path) => {
  let response
  try {
    response = await client.collection(FILES_DOCTYPE).statByPath(path)
  } catch (error) {
    if (error && error.status === 404) return null
    throw new Error(error.message)
  }

  return response.data
}

export const filesMutations = client => ({
  createDirectoryByPath: createDirectoryByPath.bind(null, client),
  statDirectoryByPath: statDirectoryByPath.bind(null, client)
})

export default filesMutations
