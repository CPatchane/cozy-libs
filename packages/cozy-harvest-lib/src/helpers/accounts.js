import get from 'lodash/get'
import merge from 'lodash/merge'

import manifest from './manifest'

const DEFAULT_TWOFA_CODE_PROVIDER_TYPE = 'default'

const TWOFA_NEEDED_STATUSES = {
  [DEFAULT_TWOFA_CODE_PROVIDER_TYPE]: 'TWOFA_NEEDED',
  email: 'TWOFA_NEEDED.EMAIL',
  sms: 'TWOFA_NEEDED.SMS'
}

const TWOFA_RETRY_STATUS = 'TWOFA_NEEDED_RETRY'
const RESET_SESSION_STATE = 'RESET_SESSION'

/**
 * Return a boolean to know if the account is in a two fa code needed
 * status
 * @param  {String}  status Account two FA Status
 * @return {Boolean}
 */
export const isTwoFANeeded = status => {
  for (let s in TWOFA_NEEDED_STATUSES) {
    if (TWOFA_NEEDED_STATUSES[s] === status) {
      return true
    }
  }
  return false
}

export const isTwoFARetry = status => {
  return status === TWOFA_RETRY_STATUS
}

/**
 * Return the status object key matching the status value
 * @param  {String} status Account document
 * @return {String}        Two FA Code providing type or default one if not known
 */
export const getTwoFACodeProvider = account => {
  if (!account || !account.state) return DEFAULT_TWOFA_CODE_PROVIDER_TYPE
  return Object.keys(TWOFA_NEEDED_STATUSES).find(
    s => TWOFA_NEEDED_STATUSES[s] === account.state
  )
}

/**
 * Returns the label for the given account.
 * This label is by default the value for the identifier field.
 * If there is no value for this field, the label is the io.cozy.accounts
 * document id.
 * @param  {Object} account io.cozy.accounts documents
 * @return {string}         The label associated to this account.
 */
export const getLabel = account =>
  get(account, `auth.${account.identifier}`) || account._id

/**
 * Transforms AccountForm data to io.cozy.accounts document
 * @param  {object} konnector Konnector related to account
 * @param  {object} data      Data from AccountForm
 * @return {object}           io.cozy.accounts attributes
 */
export const build = (konnector, authData) => {
  // We are not at the final target for io.cozy.accounts.
  // For now we are just ensuring legacy
  return {
    auth: authData,
    account_type: konnector.slug,
    identifier: manifest.getIdentifier(konnector.fields)
  }
}

/**
 * Merges existing io.cozy.accounts auth with Auth data from AccountForm
 * @param  {object} account   io.cozy.accounts document
 * @param  {object} data      Data from AccountForm
 * @return {object}           io.cozy.accounts attributes
 */
export const mergeAuth = (account, authData) => ({
  ...account,
  auth: merge(account.auth, authData)
})

/**
 * Update Two FA code from TwoFAForm into io.cozy.accounts document
 * @param  {object} account   io.cozy.accounts document
 * @param  {object} code      Code from TwoFAForm
 * @return {object}           io.cozy.accounts attributes
 */
export const updateTwoFaCode = (account, code) => ({
  ...account,
  // reset the state,
  state: null,
  twofa_code: code
})
/**
 * Set a state to reset the konnector session into io.cozy.accounts document
 * only if necessary, if password/passphrase have changed
 * @param  {object} account   io.cozy.accounts document
 * @return {object}           io.cozy.accounts updated document
 */
export const setSessionResetIfNecessary = (account, changedFields) => {
  const isPasswordChanged =
    !!account && !!(changedFields.password || changedFields.passphrase)
  return isPasswordChanged
    ? {
        ...account,
        state: RESET_SESSION_STATE
      }
    : account
}

export default {
  build,
  getLabel,
  getTwoFACodeProvider,
  isTwoFANeeded,
  isTwoFARetry,
  mergeAuth,
  setSessionResetIfNecessary,
  updateTwoFaCode
}
