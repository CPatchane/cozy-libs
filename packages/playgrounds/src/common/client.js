import CozyClient from 'cozy-client'

export default new CozyClient({
  scope: ['io.cozy.apps', 'io.cozy.konnectors'],
  schema: {
    apps: {
      doctype: 'io.cozy.apps'
    },
    konnectors: {
      doctype: 'io.cozy.konnectors'
    }
  },
  oauth: {
    clientName: 'Example App',
    softwareID: 'io.cozy.example',
    redirectURI: 'http://localhost:1234/auth'
  }
})
