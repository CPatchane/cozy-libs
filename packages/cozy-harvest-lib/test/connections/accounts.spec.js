/* eslint-env jest */

import client from 'cozy-client'

jest.mock('cozy-client', () => ({
  collection: jest.fn().mockReturnValue({
    add: jest.fn()
  }),
  create: jest.fn(),
  get: jest.fn(),
  query: jest.fn(),
  save: jest.fn(),
  where: jest.fn()
}))

const fixtures = {
  accountPermissions: {
    source_id: `io.cozy.konnectors/kaggregated`,
    permissions: {
      aggregatorAccount: {
        type: 'io.cozy.accounts',
        verbs: ['GET', 'PUT'],
        values: ['io.cozy.accounts.kaggregated-aggregator']
      }
    }
  },
  konnector: {
    slug: 'konnectest'
  },
  konnectorWithAggregator: {
    slug: 'kaggregated',
    aggregator: {
      accountId: 'kaggregated-aggregator'
    }
  },
  simpleAccount: {
    account_type: 'test',
    auth: {
      login: 'login',
      password: 'pass'
    }
  },
  parentAccount: {
    _id: 'kaggregated-aggregator'
  }
}

import { accountsMutations } from 'connections/accounts'
const { createAccount, updateAccount } = accountsMutations(client)

describe('Account mutations', () => {
  beforeAll(() => {
    client.create.mockResolvedValue({ data: fixtures.simpleAccount })
    client.save.mockResolvedValue({ data: fixtures.simpleAccount })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  describe('createAccount', () => {
    it('calls Cozy Client and return account', async () => {
      const account = await createAccount(
        fixtures.konnector,
        fixtures.simpleAccount
      )

      expect(client.create).toHaveBeenCalledWith(
        'io.cozy.accounts',
        fixtures.simpleAccount
      )
      expect(account).toEqual(fixtures.simpleAccount)
    })

    const simpleAccountFixtureWithMasterRelation = {
      ...fixtures.simpleAccount,
      relationships: {
        parent: {
          data: {
            _id: 'kaggregated-aggregator',
            _type: 'io.cozy.accounts'
          }
        }
      }
    }

    beforeEach(() => {
      client.get.mockReset()
      client.get.mockResolvedValue({
        data: simpleAccountFixtureWithMasterRelation
      })
    })

    it('throws an error when konnector have no aggregator attribute', async () => {
      const konnector = {}
      expect(createAccount(konnector, fixtures.simpleAccount)).rejects.toEqual(
        new Error('Konnector does not provide aggregator account id')
      )
    })

    it('throws an error when konnector have no aggregator.accountId attribute', async () => {
      const konnector = { aggregator: {} }
      expect(createAccount(konnector, fixtures.simpleAccount)).rejects.toEqual(
        new Error('Konnector does not provide aggregator account id')
      )
    })

    describe('if parent account does not exists', () => {
      beforeEach(() => {
        client.get = jest.fn().mockRejectedValue({ status: 404 })

        client.create.mockReset()
        client.create.mockResolvedValue({
          data: simpleAccountFixtureWithMasterRelation
        })
        client.create.mockResolvedValueOnce({ data: fixtures.parentAccount })
      })

      it('creates parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith('io.cozy.accounts', {
          _id: fixtures.konnectorWithAggregator.aggregator.accountId
        })
      })

      it('adds permissions to parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.collection().add).toHaveBeenCalledWith(
          fixtures.konnectorWithAggregator,
          fixtures.accountPermissions.permissions
        )
      })

      it('creates child account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith(
          'io.cozy.accounts',
          simpleAccountFixtureWithMasterRelation
        )
      })

      it('throws error if get fails', async () => {
        client.get.mockReset()
        client.get.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'An error occurred when finding parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if permission add fails', async () => {
        client
          .collection()
          .add.mockRejectedValueOnce(new Error('Mocked add Error'))
        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'Cannot set permission for account kaggregated-aggregator (Mocked add Error)'
          )
        )
      })

      it('returns child account', async () => {
        const account = await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(account).toEqual(simpleAccountFixtureWithMasterRelation)
      })

      it('throws error if parent account creation fails', async () => {
        client.create.mockReset()
        client.create.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'Cannot create parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if child account creation fails', async () => {
        client.create.mockReset()
        client.create.mockResolvedValueOnce({ data: fixtures.parentAccount })
        client.create.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(new Error('Mocked error'))
      })
    })

    describe('if parent account exists', () => {
      beforeEach(() => {
        client.query.mockReset()
        client.query = jest
          .fn()
          .mockResolvedValue({ data: [fixtures.parentAccount] })
        client.create.mockResolvedValue({
          data: simpleAccountFixtureWithMasterRelation
        })
      })

      it('does not create parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).not.toHaveBeenCalledWith('io.cozy.accounts', {
          _id: fixtures.konnectorWithAggregator.aggregator.accountId
        })
      })

      it('adds permissions to parent account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.collection().add).toHaveBeenCalledWith(
          fixtures.konnectorWithAggregator,
          fixtures.accountPermissions.permissions
        )
      })

      it('creates child account', async () => {
        await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(client.create).toHaveBeenCalledWith(
          'io.cozy.accounts',
          simpleAccountFixtureWithMasterRelation
        )
      })

      it('returns child account', async () => {
        const account = await createAccount(
          fixtures.konnectorWithAggregator,
          fixtures.simpleAccount
        )

        expect(account).toEqual(simpleAccountFixtureWithMasterRelation)
      })

      it('throws error if get fails', async () => {
        client.get.mockReset()
        client.get.mockRejectedValue(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(
          new Error(
            'An error occurred when finding parent account kaggregated-aggregator (Mocked error)'
          )
        )
      })

      it('throws error if child account creation fails', async () => {
        client.create.mockReset()
        client.create.mockRejectedValueOnce(new Error('Mocked error'))

        await expect(
          createAccount(
            fixtures.konnectorWithAggregator,
            fixtures.simpleAccount
          )
        ).rejects.toEqual(new Error('Mocked error'))
      })
    })
  })

  describe('updateAccount', () => {
    it('call CozyClient::save and returns account', async () => {
      const account = await updateAccount(fixtures.simpleAccount)
      expect(client.save).toHaveBeenCalledWith(fixtures.simpleAccount)
      expect(account).toEqual(fixtures.simpleAccount)
    })
  })
})
