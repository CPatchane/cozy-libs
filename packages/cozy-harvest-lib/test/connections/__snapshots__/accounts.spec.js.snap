// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account mutations watchAccount should return an account watcher with correct properties 1`] = `
KonnectorAccountWatcher {
  "account": Object {
    "accountType": "dummy",
    "auth": Object {},
  },
  "client": Object {
    "collection": [MockFunction],
    "create": [MockFunction],
    "query": [MockFunction],
    "save": [MockFunction],
    "where": [MockFunction],
  },
  "options": Object {
    "onError": [MockFunction],
    "onLoginSuccess": [MockFunction],
    "onSuccess": [MockFunction],
  },
}
`;
